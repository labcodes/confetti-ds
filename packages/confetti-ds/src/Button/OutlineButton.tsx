import React from "react";

import AbstractButton from "./AbstractButton";
import { BaseButtonProps } from "./types";

export interface OutlineButtonProps extends BaseButtonProps {
  theme?: "teal" | "purple" | "light" | "dark";
}

export default function OutlineButton({
  label,
  kind = "outline",
  type = "button",
  theme = "teal",
  hasIcon = false,
  icon,
  size = "normal",
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  fullWidth = false,
  tabIndex,
}: OutlineButtonProps) {
  return (
    <AbstractButton
      kind={kind}
      label={label}
      type={type}
      theme={theme}
      icon={icon}
      hasIcon={hasIcon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}
