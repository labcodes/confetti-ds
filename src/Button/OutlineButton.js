import React from "react";
import PropTypes from "prop-types";

import AbstractButton from "./AbstractButton";
import { ICON_TYPES } from "../constants";

export default function OutlineButton({
  text,
  type,
  skin,
  icon,
  size,
  disabled,
  onClick,
  fullWidth,
  ariaDisabled,
  tabIndex,
}) {
  return (
    <AbstractButton
      variant="outline"
      text={text}
      type={type}
      skin={skin}
      icon={icon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}

OutlineButton.propTypes = {
  /** This is the button label. */
  text: PropTypes.string.isRequired,
  /** HTML type of the Button. */
  type: PropTypes.oneOf(["button", "submit", "reset"]),
  /** Sets a special color skin to the button. */
  skin: PropTypes.oneOf(["", "light", "dark"]),
  /** Sets the icon related to the button label. Default state: no icon. */
  icon: PropTypes.oneOf(ICON_TYPES),
  /** Sets the button's height. Small = 32px, Normal = 40px, Large = 48px. */
  size: PropTypes.oneOf(["normal", "small", "large"]),
  /** Disables the Button. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled: PropTypes.bool,
  /** Disables the Button. Won't be read by screen readers. */
  disabled: PropTypes.bool,
  /** Action to be executed when the button is clicked. */
  onClick: PropTypes.func,
  /** Makes the button expand to its container's full width. */
  fullWidth: PropTypes.bool,
  /* Sets the order of which the elements will be focused on. Default value: 0. */
  tabIndex: PropTypes.string,
};

OutlineButton.defaultProps = {
  skin: undefined,
  icon: undefined,
  type: "button",
  size: "normal",
  disabled: false,
  ariaDisabled: false,
  onClick: () => {},
  fullWidth: false,
  tabIndex: undefined,
};
