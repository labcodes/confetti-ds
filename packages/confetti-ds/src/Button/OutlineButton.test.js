import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import OutlineButton from "./OutlineButton";

describe("OutlineButton", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<OutlineButton text="Test Outline Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(<OutlineButton text="Test Outline Disabled Button 1" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <OutlineButton text="Test Outline Disabled Button 2" disabled />
    );
    expect(mountedComponent.find("button").prop("disabled")).toEqual(true);
  });

  it("renders as expected when passing a light skin", async () => {
    const renderedComponent = renderer
      .create(<OutlineButton text="Test Outline Light Button" skin="light" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a large size", async () => {
    const renderedComponent = renderer
      .create(<OutlineButton text="Test Outline Large Button" size="large" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a magnifying-glass icon", async () => {
    const renderedComponent = renderer
      .create(
        <OutlineButton
          text="Test Outline magnifying-glass Button"
          icon="magnifying-glass"
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("calls props.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowButton = shallow(
      <OutlineButton text="Test Click on Outl Button" onClick={mockOnClick} />
    );
    shallowButton.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it("doesn't call onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const mountedComponent = mount(
      <OutlineButton
        ariaDisabled
        onClick={mockOnClick}
        text="Test Outl Button"
      />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    mountedComponent.find("button").simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(0);
  });

  it("renders as expected if full width", async () => {
    const renderedComponent = renderer
      .create(<OutlineButton text="Test Outline fullWidth Button" fullWidth />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with full width", async () => {
    const mountedComponent = mount(
      <OutlineButton text="Test fullWidth Button" fullWidth />
    );
    expect(mountedComponent.find(".lab-btn--block")).toHaveLength(1);
  });

  it("renders as expected when tabIndex is passed", async () => {
    const renderedComponent = renderer
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      .create(<OutlineButton text="Test tabIndex" tabIndex="1" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
