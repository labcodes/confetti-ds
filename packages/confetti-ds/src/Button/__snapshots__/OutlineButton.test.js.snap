// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutlineButton renders as expected if full width 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal lab-btn--block"
  onClick={[Function]}
  type="button"
>
  
  Test Outline fullWidth Button
</button>
`;

exports[`OutlineButton renders as expected when passing a large size 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--large"
  onClick={[Function]}
  type="button"
>
  
  Test Outline Large Button
</button>
`;

exports[`OutlineButton renders as expected when passing a light skin 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal lab-btn--light"
  onClick={[Function]}
  type="button"
>
  
  Test Outline Light Button
</button>
`;

exports[`OutlineButton renders as expected when passing a magnifying-glass icon 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  <span
    className="lab-icon lab-icon--magnifying-glass lab-icon--white lab-icon--petit lab-btn__icon"
  />
  Test Outline magnifying-glass Button
</button>
`;

exports[`OutlineButton renders as expected when passing disabled as true 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  
  Test Outline Disabled Button 1
</button>
`;

exports[`OutlineButton renders with base props 1`] = `
<button
  className="lab-btn lab-btn--outline lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Outline Button
</button>
`;
